### PR TITLE
Copy route token to clipboard

### PR DESCRIPTION
**Pull request by ChenyuHeidiZhang**
Add a button to copy the route token to clipboard

### CODE REVIEW

File: custom-routes-vis/src/main/webapp/script.js [hunk 1]
```diff
@@ -335,15 +335,56 @@ function selectRouteDisplayDetails(
...
+      copyTokenToClipBoard(routeToken);
+    });
+    routeInfoElement.appendChild(copyBtn);
+
     console.log('Route token:', routeToken);
     updateDeepLinkingUrl(routeToken);
   }
 }
 
+/** Copies the given route token to the clipboard. */
```
Comment (carmenBen) [inline custom-routes-vis/src/main/webapp/script.js:353]: Add a param doc.

File: custom-routes-vis/src/main/webapp/script.js [hunk 2]
```diff
@@ -335,15 +335,56 @@ function selectRouteDisplayDetails(
...
+    alert('Successfully copied to clipboard!');
+  }, function(err) {
+    alert('Could not copy to clipboard.');
+    console.error('Async: Could not copy token', err);
+  });
+}
+
+/** Creates a textarea for the given token and copies it to clipboard. */
+function fallbackCopyTokenToClipboard(token) {
+  var textArea = document.createElement('textarea');
```
Comment (carmenBen) [inline custom-routes-vis/src/main/webapp/script.js:370]: Change to let.

File: custom-routes-vis/src/main/webapp/script.js [hunk 3]
```diff
@@ -335,15 +335,56 @@ function selectRouteDisplayDetails(
...
+  textArea.value = token;
+  document.body.appendChild(textArea);
+  textArea.focus();
+  textArea.select();
+
+  try {
+    var successful = document.execCommand('copy');
+    var msg = successful ? 'successful' : 'unsuccessful';
```
Comment (carmenBen) [inline custom-routes-vis/src/main/webapp/script.js:378]: Change to let.

File: custom-routes-vis/src/main/webapp/script.js [hunk 4]
```diff
@@ -335,15 +335,56 @@ function selectRouteDisplayDetails(
...
+      '\nDistance: ' + formatDistance(totalDistanceMeters) + '\n';
 
   if (routeToken !== '') {
-    routeInfoElement.innerText += '\nRoute Token: check the developer console';
+    let copyBtn = document.createElement('button');
+    copyBtn.innerText = 'Copy Route Token';
+    copyBtn.addEventListener('click', function() {
+      copyTokenToClipBoard(routeToken);
+    });
+    routeInfoElement.appendChild(copyBtn);
```
Comment (carmenBen) [inline custom-routes-vis/src/main/webapp/script.js:346]: Double checking that if the users generates routes multiple times the button is replaced each time, and they don't get like 50 buttons?

Have you considered creating a button and just hiding it or showing is as necessary?

Comment (tzhong92) [reply custom-routes-vis/src/main/webapp/script.js:346]: +1 about hiding the button if there is no route token.

Comment (ChenyuHeidiZhang) [reply custom-routes-vis/src/main/webapp/script.js:346]: Yes the button is replaced when another route is generated and yes the button is hidden if there is no route token.

File: custom-routes-vis/src/main/webapp/script.js [hunk 5]
```diff
@@ -335,15 +335,56 @@ function selectRouteDisplayDetails(
   let routeInfoElement = document.getElementById('route-info');
   routeInfoElement.innerText = 'Selected Route Info:' +
       '\nDuration: ' + formatDuration(totalDurationSec) +
-      '\nDistance: ' + formatDistance(totalDistanceMeters);
+      '\nDistance: ' + formatDistance(totalDistanceMeters) + '\n';
 
   if (routeToken !== '') {
-    routeInfoElement.innerText += '\nRoute Token: check the developer console';
+    let copyBtn = document.createElement('button');
```
Comment (hayleejane3) [inline custom-routes-vis/src/main/webapp/script.js:341]: Full words please.
(replace everywhere)
```suggestion
    let copyButton = document.createElement('button');
```

File: custom-routes-vis/src/main/webapp/script.js [hunk 6]
```diff
@@ -320,7 +320,7 @@ function createRouteFromCoordinates(
 
 /** 
  * Highlights the selected route and displays its formatted duration and 
- * distance in miles.
+ * distance in miles. Displays a copy-route-token button if applicable.
```
Comment (hayleejane3) [inline custom-routes-vis/src/main/webapp/script.js:323]: You want to say what the user will see.
```suggestion
 * distance in miles. Displays a "Copy Route Token" button if applicable.
```

File: custom-routes-vis/src/main/webapp/script.js (hunk 3 above)
Comment (hayleejane3) [inline custom-routes-vis/src/main/webapp/script.js:378]: 'message'

File: custom-routes-vis/src/main/webapp/script.js [hunk 7]
```diff
@@ -335,15 +335,56 @@ function selectRouteDisplayDetails(
...
+/** Creates a textarea for the given token and copies it to clipboard. */
+function fallbackCopyTokenToClipboard(token) {
+  var textArea = document.createElement('textarea');
+  textArea.value = token;
+  document.body.appendChild(textArea);
+  textArea.focus();
+  textArea.select();
+
+  try {
+    var successful = document.execCommand('copy');
```
Comment (hayleejane3) [inline custom-routes-vis/src/main/webapp/script.js:377]: Please keep variable names relevant and descriptive.
```suggestion
    var copyCommandStatus = document.execCommand('copy');
```

File: custom-routes-vis/src/main/webapp/script.js [hunk 8]
```diff
@@ -335,15 +335,56 @@ function selectRouteDisplayDetails(
...
+  }
+  navigator.clipboard.writeText(token).then(function() {
+    alert('Successfully copied to clipboard!');
+  }, function(err) {
+    alert('Could not copy to clipboard.');
+    console.error('Async: Could not copy token', err);
+  });
+}
+
+/** Creates a textarea for the given token and copies it to clipboard. */
```
Comment (ntarn) [inline custom-routes-vis/src/main/webapp/script.js:368]: param doc here as well

File: custom-routes-vis/src/main/webapp/script.js (hunk 7 above)
Comment (ntarn) [inline custom-routes-vis/src/main/webapp/script.js:377]: also change to let here

File: custom-routes-vis/src/main/webapp/script.js (hunk 6 above)
Comment (ChenyuHeidiZhang) [reply custom-routes-vis/src/main/webapp/script.js:323]: ok

File: custom-routes-vis/src/main/webapp/script.js (hunk 7 above)
Comment (ChenyuHeidiZhang) [reply custom-routes-vis/src/main/webapp/script.js:377]: ok

**Pull request merged**